docs(upgrade): add/update docs

gkalpak opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?
```
[x] Documentation content changes
```

## Other information
Includes a commit from #24583 to get the updated AngularJS typings.

gkalpak committed:
docs(upgrade): add API docs for `propagateDigest`

gkalpak committed:
docs(upgrade): add API docs for `downgradeModule()`

gkalpak committed:
docs: minor fixes in `docs-style-guide`

gkalpak committed:
docs(upgrade): update API docs for upgrade/static

mary-poppins commented:
You can preview 011b033 at https://pr18487-011b033.ngbuilds.io/.

gkalpak committed:
test: minor improvements in examples e2e tests script

mary-poppins commented:
You can preview 252185d at https://pr18487-252185d.ngbuilds.io/.

gkalpak committed:
docs(upgrade): add guide about `downgradeModule()`

mary-poppins commented:
You can preview 289913c at https://pr18487-289913c.ngbuilds.io/.

mary-poppins commented:
You can preview 5e7e393 at https://pr18487-5e7e393.ngbuilds.io/.

mary-poppins commented:
You can preview af94c36 at https://pr18487-af94c36.ngbuilds.io/.

mhevery commented on packages/upgrade/src/static/upgrade_module.ts:
```diff
@@ -65,33 +75,50 @@ import {NgAdapterInjector} from './util';
  *    a. This implies that the component bindings will always follow the semantics of the
```
the formatting is messed up here. https://pr18487-af94c36.ngbuilds.io/api/upgrade/static/UpgradeModule

petebacondarwin replied:
We must always use `1.`, `2.` etc, to indicate ordered lists, even for sub-lists.
We can change the sublist to display as `a.`, `b.` etc, via CSS.

mhevery commented on packages/upgrade/src/static/upgrade_module.ts:
```diff
@@ -65,33 +75,50 @@ import {NgAdapterInjector} from './util';
+ * ### Upgrading an AngularJS service
  *
  * There is no specific API for upgrading an AngularJS service. Instead you should just follow the
  * following recipe:
```
I am not a fan of the example below since it uses a string as the injectable. `@Inject('titleCase')` which makes it seem that all of the Angular / AngularJS must use string.

Could we change it to something which could use a class?

```
class Greeter {...}

ng1Module.service('greeter', Greeter);

{provide: Greeter, useFactory: (i) => i.get('greeter'), deps: ['$injector']}

constructor(greeter: Greeter) {...}
```

I think this example is better because it shows how a `'greeter'` string injection token becomes just a `Greeter` type which is than injected without any additional text.

petebacondarwin replied:
Fixed.

mhevery commented on packages/upgrade/src/static/downgrade_module.ts:
```diff
@@ -17,7 +17,93 @@ import {angular1Providers, setTempInjectorRef} from './angular1_providers';
+ * Allows an Angular module to be bootstrapped "on-demand" (possibly lazily) from within a running
+ * AngularJS application.
+ *
+ * @howToUse
```
The example relies on JIT. Given that we are moving to AoT could the example be AoT? (get rid of JIT)

gkalpak replied:
According to my understanding, putting together the infrastructure to run examples in AoT mode, is something that needs to happen for all examples, so doesn't have to be part of this PR imo.

mhevery commented on packages/upgrade/src/static/downgrade_module.ts:
```diff
@@ -17,7 +17,93 @@ import {angular1Providers, setTempInjectorRef} from './angular1_providers';
+ * services) will not be available until the downgraded module has been bootstrapped, i.e. by
+ * instantiating a downgraded component.
+ *
+ * ### Differences with `UpgradeModule`
```
maybe move this section to a guide which would help to compare contrast the two approaches? 
Also there should be a link from here to the guide.

gkalpak replied:
I have added a link to the guide, which discusses the differences in more detail. I think it is still useful to have this brief overview here.

mhevery commented on aio/content/guide/upgrade-lazy.md:
```diff
@@ -0,0 +1,357 @@
+# Upgrading from AngularJS (in a more flexible way)
+
+<div class="alert is-helpful">
```
I would like to see some discussion on `UpgradeModule` puts Angular at the root, where as `downgradeModule` put AngularJS at the root.

gkalpak replied:
I am not sure what you mean by "at the root". This could be true in the sense that either Angular or AngularJS owns the "root of the template" and the other framework can be restricted to only part of the app.

While this is true for `downgradeModule()` and theoretically possible for `UpgradeModule`, we've been telling people that AngularJS must own the root of the template with `UpgradeModule` (even if Anglar is bootstrapped first). Also note, that currently it is not possible to load AngularJS lazily (I mean it is technically possible, but currently not implemented).

So, happy to add something about Angular/AngularJS being at the root, but I don't know what exactly to say :smiley:

petebacondarwin committed:
docs(upgrade): fix sub-ordered-list syntax

We must always use 1., 2. etc, to indicate ordered lists, even for sub-lists.
We can change the sublist to display as a., b. etc, via CSS.

petebacondarwin committed:
docs(upgrade): use a class for upgraded service

This makes the resulting use in Angular more ideomatic, since we can just
use the class type as the injection indicator.

petebacondarwin commented on packages/upgrade/src/static/downgrade_module.ts:
```diff
@@ -17,7 +17,93 @@ import {angular1Providers, setTempInjectorRef} from './angular1_providers';
+ * Allows an Angular module to be bootstrapped "on-demand" (possibly lazily) from within a running
+ * AngularJS application.
+ *
+ * @howToUse
```
Do we have the facility to run AoT on our examples?
If not, would this have to be done manually and the resulting NgModuleFactory file committed?

mary-poppins commented:
You can preview 8c7e5d5 at https://pr18487-8c7e5d5.ngbuilds.io/.

mary-poppins commented:
You can preview da9a812 at https://pr18487-da9a812.ngbuilds.io/.

kapunahelewong committed:
docs(aio): tech edits to upgrade-lazy

mary-poppins commented:
You can preview 7c0e363 at https://pr18487-7c0e363.ngbuilds.io/.

mary-poppins commented:
You can preview 456e3e7 at https://pr18487-456e3e7.ngbuilds.io/.

mary-poppins commented:
You can preview b4d25c4 at https://pr18487-b4d25c4.ngbuilds.io/.

mary-poppins commented:
You can preview aedf0aa at https://pr18487-aedf0aa.ngbuilds.io/.

mary-poppins commented:
You can preview 08f89cf at https://pr18487-08f89cf.ngbuilds.io/.

mary-poppins commented:
You can preview 47d394f at https://pr18487-47d394f.ngbuilds.io/.

gkalpak committed:
fixup! docs(upgrade): add API docs for `downgradeModule()`

mary-poppins commented:
You can preview dcec707 at https://pr18487-dcec707.ngbuilds.io/.

mary-poppins commented:
You can preview 810d8ac at https://pr18487-810d8ac.ngbuilds.io/.

mhevery commented:
What is preventing this from getting merged?

gkalpak commented:
Mainly the fact that building examples was failing with the TypeScript version we use now due to bad AngularJS typings (https://github.com/angular/angular/pull/18487#issuecomment-330636786).
I never got around to fixing the typings.

##
We still plan to ship our own typings for AngularJS, but probably after the 1.7.0 release. It might be better to land this even if we have to add some `any` here and there to get around that.
I'll try to get it to a mergeable state next week.

sharikovvladislav commented:
Just do it!

mary-poppins commented:
You can preview 48ca924 at https://pr18487-48ca924.ngbuilds.io/.

mary-poppins commented:
You can preview 76676f8 at https://pr18487-76676f8.ngbuilds.io/.

mary-poppins commented:
You can preview 1655d21 at https://pr18487-1655d21.ngbuilds.io/.

mary-poppins commented:
You can preview dff7c04 at https://pr18487-dff7c04.ngbuilds.io/.

mary-poppins commented:
You can preview 88da92c at https://pr18487-88da92c.ngbuilds.io/.

mary-poppins commented:
You can preview 5060c97 at https://pr18487-5060c97.ngbuilds.io/.

mary-poppins commented:
You can preview 1320a87 at https://pr18487-1320a87.ngbuilds.io/.

mary-poppins commented:
You can preview 192a114 at https://pr18487-192a114.ngbuilds.io/.

mary-poppins commented:
You can preview b24ab73 at https://pr18487-b24ab73.ngbuilds.io/.

mary-poppins commented:
You can preview 871b4db at https://pr18487-871b4db.ngbuilds.io/.

mhevery commented:
http://test/OCL:201582831:BASE:201582848:1529615168946:1d3287f6

mary-poppins commented:
You can preview 7a6fdfa at https://pr18487-7a6fdfa.ngbuilds.io/.

mary-poppins commented:
You can preview f1ed719 at https://pr18487-f1ed719.ngbuilds.io/.

mhevery commented:
@gkalpak Argh there is a conflict, can you resolve?

gkalpak commented:
@mhevery, resolved.

mary-poppins commented:
You can preview 2262b75 at https://pr18487-2262b75.ngbuilds.io/.

gkalpak committed:
refactor(upgrade): fix examples for `strictPropertyInitialization` and remove internal comments

gkalpak commented:
:tada: :tada: :tada:
